compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEUnrollAndDistributeWorkgroup.cpp

```diff
@@ -0,0 +1,503 @@
+                 dyn_cast<IntegerAttr>(sourceMemspace).getInt()) {
+        // More locally, towards AIE
+        internalLogicalObjectFifos.insert(targetObjectFifo);
+      } else if (dyn_cast<IntegerAttr>(targetMemspace).getInt() <
+                 dyn_cast<IntegerAttr>(sourceMemspace).getInt()) {
+        // More globally, away from AIE
+        internalLogicalObjectFifos.insert(sourceObjectFifo);
+      }
+      return WalkResult::advance();
+    });
+  }
+
+  /// Hoist dma ops outside the scf.for operation if none of the dma op's
+  /// operands depend on the scf.for's induction variable or any operation
+  /// within the scf.for's scope.
+  LogicalResult hoistDmaOps(PatternRewriter &rewriter, scf::ForOp forOp) const {
+    bool hoistHappened{false};
+
+    // Create utility function to check whether an operand depends on a scf.for
+    // induction variable or a value within the scf.for's scope
+    auto dependsOnLoop = [&](OpOperand &operand) -> bool {
+      Operation *op = operand.get().getDefiningOp();
+      if (!op) return operand.get() == forOp.getInductionVar();
+
+      // Check fo induction var and whether the parent scf.for is the same as
+      // the one we're using for the hoist
+      auto parentForOp = op->getParentOfType<scf::ForOp>();
+      return (operand.get() == forOp.getInductionVar()) ||
+             (parentForOp && parentForOp == forOp);
+    };
+
+    // Collect the internal logical objectfifos based on dma ops' sources and
+    // targets. Then walk all dma ops and try to hoist them either before or
+    // after the scf.for depending on whether target or source logical
+    // objectfifo is an internal one.
+    DenseSet<AMDAIE::LogicalObjectFifoFromMemrefOp> internalLogicalObjectFifos;
+    collectInternalLogicalObjectFifos(forOp, internalLogicalObjectFifos);
+    forOp.walk([&](AMDAIE::DmaCpyNdOp dmaOp) {
+      if ((!internalLogicalObjectFifos.contains(dmaOp.getSourceObjectFifo()) ||
+           !internalLogicalObjectFifos.contains(dmaOp.getTargetObjectFifo())) &&
+          !llvm::any_of(dmaOp->getOpOperands(), [&](OpOperand &operand) {
+            return dependsOnLoop(operand);
+          })) {
+        if (internalLogicalObjectFifos.contains(dmaOp.getTargetObjectFifo()))
+          rewriter.moveOpBefore(dmaOp, forOp);
+        if (internalLogicalObjectFifos.contains(dmaOp.getSourceObjectFifo()))
+          rewriter.moveOpAfter(dmaOp, forOp);
+        hoistHappened = true;
+      }
+      return WalkResult::advance();
+    });
+    if (hoistHappened) return success();
+    return failure();
+  }
+
+  /// Unroll the for loop and while doing so, create new clones of
+  /// AMDAIE::LogicalObjectFifoFromMemrefOp so they can be distributed
+  /// onto multiple physical locations later.
+  LogicalResult loopUnrollAndDistributeLogicalObjectFifos(
+      RewriterBase &rewriter, scf::ForOp forOp) const {
+    Block *loopBodyBlock = forOp.getBody();
+    auto builder = OpBuilder::atBlockTerminator(loopBodyBlock);
```

Comment:
Expand `auto` to explicit types here and elsewhere in the code.

We should use `auto` only when the type can be seen from the in-line context. Cases where `auto` is favourable are `cast<SOME_TYPE>`, `dyn_cast<SOME_TYPE>`, `rewriter.create<SOME_OP>`, etc.

Reply:
Done.

Reply:
Although, I don't think this is practiced very carefully throughout iree. I am myself more on the 'use more auto side', especially for simple/common types, but I am ok with being more strict in this codebase as long as it's consistent.

Reply:
You're correct. Just that I came across this coding style practice during one of the upstream PRs in `llvm-project`  recently and thus trying to adopt it here too. It definitely helps.

Thanks for following up!